lib/Dialect/SV/Transforms/HWSpecialize.cpp

```diff
@@ -0,0 +1,298 @@
+
+  LogicalResult matchAndRewrite(ParamValueOp op,
+                                PatternRewriter &rewriter) const override {
+    // Substitute the param value op with an evaluated constant operation.
+    auto paramValue =
+        evaluateParametricAttr(op.getLoc(), parameters, op.value());
+    if (failed(paramValue))
+      return failure();
+    rewriter.replaceOpWithNewOp<hw::ConstantOp>(
+        op, op.getType(), paramValue.getValue().getSExtValue());
+    return success();
+  }
+
+  ArrayAttr parameters;
+};
+
+// hw.array_get operations require indexes to be of equal width of the
+// array itself. Since indexes may originate from constants or parameters, emit
+// comb.extract operations to fulfil this invariant.
+struct NarrowArrayGetIndexPattern : public OpConversionPattern<ArrayGetOp> {
+public:
+  using OpConversionPattern<ArrayGetOp>::OpConversionPattern;
+
+  LogicalResult
+  matchAndRewrite(ArrayGetOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    auto inputType = type_cast<ArrayType>(op.input().getType());
+    Type targetIndexType = IntegerType::get(
+        getContext(),
+        inputType.getSize() == 1 ? 1 : llvm::Log2_64_Ceil(inputType.getSize()));
+
+    if (op.index().getType().getIntOrFloatBitWidth() ==
+        targetIndexType.getIntOrFloatBitWidth())
+      return failure(); // nothing to do
+
+    // Narrow the index value.
+    auto narrowedIndex =
+        narrowValueToArrayWidth(rewriter, op.input(), op.index());
+    rewriter.replaceOpWithNewOp<ArrayGetOp>(op, op.input(), narrowedIndex);
+    return success();
+  }
+};
+
+// Generic pattern to convert parametric result types.
+struct ParametricTypeConversionPattern : public ConversionPattern {
+  ParametricTypeConversionPattern(MLIRContext *ctx,
+                                  TypeConverter &typeConverter,
+                                  ArrayAttr parameters)
+      : ConversionPattern(typeConverter, MatchAnyOpTypeTag(), /*benefit=*/1,
+                          ctx),
+        parameters(parameters) {}
+
+  LogicalResult
+  matchAndRewrite(Operation *op, ArrayRef<Value> operands,
+                  ConversionPatternRewriter &rewriter) const override {
+    llvm::SmallVector<Value, 4> convertedOperands;
+    // Update the result types of the operation
+    bool ok = true;
+    rewriter.updateRootInPlace(op, [&]() {
+      // Mutate result types
+      for (auto &it : llvm::enumerate(op->getResultTypes())) {
+        auto res = evaluateParametricType(op->getLoc(), parameters, it.value());
+        ok &= succeeded(res);
+        if (!ok)
+          return;
+        op->getResult(it.index()).setType(res.getValue());
+      }
+    });
+
+    // Note: 'operands' have already been converted with the supplied type
+    // converter to this pattern. Make sure that we materialize this conversion
+    // by updating the operands to op.
+    op->setOperands(operands);
```

Comment:
I'm pretty sure this also needs to be inside the `updateRootInPlace(op, ...)`, since you're mutating `op`.